tests/test_updater_fetch_target.py

```diff
@@ -0,0 +1,166 @@
+
+"""Test 'Fetch target' from 'Detailed client workflow' as well as
+target files storing/loading from cache.
+"""
+import os
+import sys
+import tempfile
+import unittest
+from typing import Optional, Tuple
+
+from tests import utils
+from tests.repository_simulator import RepositorySimulator
+from tuf.exceptions import RepositoryError
+from tuf.ngclient import Updater
+
+
+class TestFetchTarget(unittest.TestCase):
+    """Test ngclient downloading and caching target files."""
+
+    # set dump_dir to trigger repository state dumps
+    dump_dir: Optional[str] = None
+
+    def setUp(self) -> None:
+        # pylint: disable-next=consider-using-with
+        self.temp_dir = tempfile.TemporaryDirectory()
+        self.metadata_dir = os.path.join(self.temp_dir.name, "metadata")
+        self.targets_dir = os.path.join(self.temp_dir.name, "targets")
+        os.mkdir(self.metadata_dir)
+        os.mkdir(self.targets_dir)
+
+        # Setup the repository, bootstrap client root.json
+        self.sim = RepositorySimulator()
+        with open(os.path.join(self.metadata_dir, "root.json"), "bw") as f:
+            f.write(self.sim.signed_roots[0])
+
+        if self.dump_dir is not None:
+            # create test specific dump directory
+            name = self.id().split(".")[-1]
+            self.sim.dump_dir = os.path.join(self.dump_dir, name)
+            os.mkdir(self.sim.dump_dir)
+
+    def tearDown(self) -> None:
+        self.temp_dir.cleanup()
+
+    def _init_updater(self) -> Updater:
+        """Creates a new updater and runs refresh."""
+        if self.sim.dump_dir is not None:
+            self.sim.write()
+
+        updater = Updater(
+            self.metadata_dir,
+            "https://example.com/metadata/",
+            self.targets_dir,
+            "https://example.com/targets/",
+            self.sim,
+        )
+        return updater
+
+    targets: utils.DataSet = {
+        "standard case": ("targetpath", b"content", "targetpath"),
+        "non-asci case": ("åäö", b"more content", "%C3%A5%C3%A4%C3%B6"),
+        "subdirectory case": (
+            "a/b/c/targetpath",
+            b"dir target content",
+            "a%2Fb%2Fc%2Ftargetpath",
+        ),
+    }
+
+    @utils.run_sub_tests_with_dataset(targets)
+    def test_fetch_target(self, test_case_data: Tuple[str, bytes, str]) -> None:
+        targetpath, content, encoded_path = test_case_data
+        path = os.path.join(self.targets_dir, encoded_path)
+
+        updater = self._init_updater()
+        # target does not exist yet
+        self.assertIsNone(updater.get_targetinfo(targetpath))
+
+        # Add targets to repository
+        self.sim.targets.version += 1
+        self.sim.add_target("targets", content, targetpath)
+        self.sim.update_snapshot()
+
+        updater = self._init_updater()
+        # target now exists, is not in cache yet
+        info = updater.get_targetinfo(targetpath)
+        assert info is not None
```

Comment:
```suggestion
        self.assertIsNotNone(info)
```

Reply:
I think these are for mypy: otherwise you'd need something like 

```
if info is None:
    self.fail("unexpected get_targetinfo() failure")
else:
    # test code using info
```

`assertIsNotNone` is not understood by mypy but `assert` is (or rather unittest): then you can avoid the indentation for the rest of the test code

Reply:
What Jussi sad :)